compiler/rustc_parse/src/validate_attr.rs

```diff
@@ -167,7 +167,20 @@ pub fn check_attribute_safety(psess: &ParseSess, safety: AttributeSafety, attr:
             // square bracket respectively.
             let diag_span = attr_item.span();
 
-            if attr.span.at_least_rust_2024() {
+            // Attributes that were safe before edition 2024, but are unsafe afterwards.
+            const SAFE_BEFORE_RUST_2024: &[Symbol] =
+                &[sym::export_name, sym::link_section, sym::no_mangle];
```

Comment:
based on https://github.com/rust-lang/rust/blob/master/compiler/rustc_feature/src/builtin_attrs.rs, I believe these are the only 3 attributes that changed their unsafety in the 2024 edition and were stable at that point.